pkg/cloud/services/networking/securitygroups.go

```diff
@@ -399,6 +402,67 @@ func (s *Service) createRule(r infrav1.SecurityGroupRule) (infrav1.SecurityGroup
 	return convertOSSecGroupRuleToConfigSecGroupRule(*rule), nil
 }
 
+func contains(list []string, element string) bool {
+	for _, e := range list {
+		if e == element {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *Service) ReconcileSecurityGroupsToInstanceAttachedInterfaces(logger logr.Logger, openStackMachine *infrav1.OpenStackMachine, attachedInterfaces []attachinterfaces.Interface, desiredSecurityGroupIDs []string) error {
+	for _, iface := range attachedInterfaces {
+		portUpdateNeeded := false
+		portID := iface.PortID
+		port, err := s.client.GetPort(portID)
+		if err != nil {
+			return fmt.Errorf("error getting port %s: %v", portID, err)
+		}
+
+		// check if port already has desired security groups
+		for _, securityGroupID := range desiredSecurityGroupIDs {
+			if !contains(port.SecurityGroups, securityGroupID) {
+				logger.Info("Port doesn't have desired security group", "portID", portID, "securityGroupID", securityGroupID)
+				portUpdateNeeded = true
+				break
+			}
+		}
+		// check if port has security groups that are not desired
+		for _, securityGroupID := range port.SecurityGroups {
+			if !contains(desiredSecurityGroupIDs, securityGroupID) {
+				logger.Info("Port has security group that is not desired", "portID", portID, "securityGroupID", securityGroupID)
+				portUpdateNeeded = true
+				break
+			}
+		}
+
+		if portUpdateNeeded {
+			logger.Info("Updating port", "portID", portID, "securityGroups", desiredSecurityGroupIDs)
+			portOpts := ports.UpdateOpts{
+				SecurityGroups: &desiredSecurityGroupIDs,
+			}
+			_, err := s.client.UpdatePort(portID, portOpts)
+			if err != nil {
+				return fmt.Errorf("error updating port %s: %v", portID, err)
+			}
+		}
+	}
+
+	appliedSecurityGroups := make([]infrav1.SecurityGroup, len(desiredSecurityGroupIDs))
+	for i, securityGroupID := range desiredSecurityGroupIDs {
+		securityGroup, err := s.client.GetSecGroup(securityGroupID)
```

Comment:
Note for myself: I could probably get rid of this call if we directly pass `[]infrav1.SecurityGroup` to this function instead of the security groups IDs that we desire. I'll take a look.